File: book/fridgets/java/fridgets/src/fridgets/FrFlow.java
package fridgets;

import java.awt.Dimension;
import java.util.Collection;
import java.util.Optional;
import nz.sodium.*;

public class FrFlow extends Fridget {
    public enum Direction { HORIZONTAL, VERTICAL };

    public FrFlow(Direction dir, Collection<Fridget> fridgets) {
        super((size, sMouse, sKey, focus, idSupply) -> {
            Cell<Dimension> desiredSize = new Cell<>(new Dimension(0,0));
            Cell<Drawable> drawable = new Cell<>(new Drawable());
            Stream<Long> sChangeFocus = new Stream<Long>();
            for (Fridget fridget : fridgets) {
                CellLoop<Optional<Dimension>> childSz = new CellLoop<>();
                Fridget.Output fo = new FrTranslate(fridget,
                  dir == Direction.HORIZONTAL
                    ? desiredSize.map(dsz -> new Dimension(dsz.width, 0))
                    : desiredSize.map(dsz -> new Dimension(0, dsz.height)))
                    .reify(childSz, sMouse, sKey, focus,
                        idSupply.child1());
                idSupply = idSupply.child2();
                childSz.loop(
                    size.lift(fo.desiredSize, (osz, foDsz) ->
                        osz.isPresent()
                            ? Optional.of(dir == Direction.HORIZONTAL
                                ? new Dimension(foDsz.width,
                                                osz.get().height)
                                : new Dimension(osz.get().width,
                                                foDsz.height))
                            : Optional.empty()
                    )
                );
                desiredSize = desiredSize.lift(fo.desiredSize,
                    dir == Direction.HORIZONTAL
                        ? (dsz, foDsz) -> new Dimension(
                            dsz.width + foDsz.width,
                            dsz.height > foDsz.height ? dsz.height
                                                      : foDsz.height)
                        : (dsz, foDsz) -> new Dimension(
                            dsz.width > foDsz.width ? dsz.width
                                                    : foDsz.width,
                            dsz.height + foDsz.height));
                drawable = drawable.lift(fo.drawable,
                    (drA, drB) -> drA.append(drB));
                sChangeFocus = sChangeFocus.orElse(fo.sChangeFocus);
            }
            return new Fridget.Output(drawable, desiredSize, sChangeFocus);
        });
    }
}

File: book/fridgets/c#/Fridgets/FrFlow.cs
using System;
using System.Collections.Generic;
using System.Windows;
using System.Windows.Controls;
using Sodium.Frp;
using Sodium.Functional;

namespace Fridgets
{
    public class FrFlow : Fridget
    {
        public FrFlow(Orientation dir, IEnumerable<Fridget> fridgets)
            : base((size, sMouse, sKey, focus, idSupply) =>
            {
                Cell<Size> desiredSize = Cell.Constant(new Size(0, 0));
                Cell<DrawableDelegate> drawable = Cell.Constant(new DrawableDelegate(d => { }));
                Stream<long> sChangeFocus = Stream.Never<long>();
                foreach (Fridget fridget in fridgets)
                {
                    CellLoop<Maybe<Size>> childSz = new CellLoop<Maybe<Size>>();
                    Output fo = new FrTranslate(fridget,
                        dir == Orientation.Horizontal
                            ? desiredSize.Map(dsz => new Point(dsz.Width, 0))
                            : desiredSize.Map(dsz => new Point(0, dsz.Height)))
                        .Reify(childSz, sMouse, sKey, focus,
                            idSupply.Child1());
                    idSupply = idSupply.Child2();
                    childSz.Loop(
                        size.Lift(fo.DesiredSize, (mSize, foDsz) =>
                            mSize.Map(s => dir == Orientation.Horizontal
                                ? new Size(foDsz.Width, s.Height)
                                : new Size(s.Width, foDsz.Height))));
                    Size LiftDesiredSizeHorizontal(Size dsz, Size foDsz) => new Size(dsz.Width + foDsz.Width, dsz.Height > foDsz.Height ? dsz.Height : foDsz.Height);
                    Size LiftDesiredSizeVertical(Size dsz, Size foDsz) => new Size(dsz.Width > foDsz.Width ? dsz.Width : foDsz.Width, dsz.Height + foDsz.Height);
                    desiredSize = desiredSize.Lift(fo.DesiredSize, dir == Orientation.Horizontal ? LiftDesiredSizeHorizontal : (Func<Size, Size, Size>)LiftDesiredSizeVertical);
                    drawable = drawable.Lift(fo.Drawable, (drA, drB) => drA.Append(drB));
                    sChangeFocus = sChangeFocus.OrElse(fo.SChangeFocus);
                }
                return new Output(drawable, desiredSize, sChangeFocus);
            })
        {
        }
    }
}